components/VM_Arm/src/modules/virtio_con.c

```diff
@@ -22,61 +42,160 @@
 #include <platsupport/serial.h>
 #include <virtioarm/virtio_console.h>
 
-static virtio_con_t *virtio_con = NULL;
-extern void *serial_getchar_buf;
+#include <virtqueue.h>
+#include <camkes/virtqueue.h>
 
-//this matches the size of the buffer in serial server
-#define BUFSIZE 4088
+/* Number of ports is the number of crossvm connections plus hvc0 */
+#define NUM_PORTS (ARRAY_SIZE(serial_layout) + 1)
 
-char txbuf[BUFSIZE];
+static virtio_con_t *virtio_con = NULL;
 extern vmm_pci_space_t *pci;
 extern vmm_io_port_list_t *io_ports;
 
-struct {
+/* 4088 because the serial_shmem_t struct has to be 0x1000 bytes big */
+#define BUFSIZE (0x1000 - 2 * sizeof(uint32_t))
+
+/**
+ * This struct is a ring buffer that occupies exactly one page, it should represent
+ * the data in the shmem region (that is shared between two nodes of a virtio console
+ * connection.
+ *
+ * The access can be lock-free since an instance of this struct can only be either the
+ * receive buffer or the sent buffer of a virtio console node, which guarantees that only
+ * one end of the ring buffer struct can be modified in a node.
+
+ * Way of indexing: `head` and `tail` should not increase beyond (BUFSIZE - 1), as
+ * they repersent the real index in `buf`.
+ */
+typedef struct serial_shmem {
     uint32_t head;
     uint32_t tail;
     char buf[BUFSIZE];
-} extern volatile *batch_buf;
+} serial_shmem_t;
+compile_time_assert(serial_shmem_1k_size, sizeof(serial_shmem_t) == 0x1000);
+
+/*
+ * Represents the virtqueues used in a given link between
+ * two VMs.
+ */
+typedef struct serial_conn {
+    void (*notify)(void);
+    volatile serial_shmem_t *recv_buf;
+    volatile serial_shmem_t *send_buf;
+} serial_conn_t;
+static serial_conn_t connections[NUM_PORTS];
+
+/* defined in virtio_console_emul.c */
+void virtio_console_putchar(int port, virtio_emul_t *con, char *buffer, uint32_t head, uint32_t tail) WEAK;
+
+/**
+ * Writes data to the guest console attached to a port, and sets the head of the receive
+ * ring buffer to the right place.
+ *
+ * @param port Port number
+ */
+static void virtio_con_notify_recv(int port)
+{
+    volatile serial_shmem_t *buffer = connections[port].recv_buf;
```

Comment:
Could add a `assert((port >=0) && (port < ARRAY_SIZE(connections)));` here. Or simply pass a `serial_conn_t*` here instead of the port type, as thgis seems to be the context we are operating on. That avoids unnecessary dependencies on the global variables.

Reply:
Ah ok, we pass the port downstream again. But it's a bit odd, shouldn't there be a cookie then for the lower layer as reference, where we don't make a strong assumption what context this holds (might still be the plain port).

Reply:
Is it necessary since the emul layer also has knowledge about the port structure?